File: android/app/src/main/kotlin/co/sumit/harbor/AndroidExitDiagnostics.kt
package co.sumit.harbor

import android.content.Context
import android.content.SharedPreferences
import java.security.MessageDigest
import java.util.concurrent.Executors

internal data class HistoricalExitRecord(
  val reason: Int,
  val status: Int,
  val importance: Int,
  val timestamp: Long
)

internal object AndroidStartupPhases {
  const val NATIVE_ON_CREATE = "native_on_create"
  private val allowed = setOf(
    NATIVE_ON_CREATE,
    "dart_main",
    "runApp",
    "first_frame",
    "database_open_started",
    "database_ready",
    "credentials_loaded",
    "binding_started",
    "binding_settled",
    "main_screen"
  )

  fun sanitize(raw: String?): String? = raw?.takeIf(allowed::contains)
}

internal class StartupPhaseStore(
  readPhase: () -> String?,
  private val persistPhase: (String) -> Boolean
) {
  val previousPhase: String? = AndroidStartupPhases.sanitize(readPhase())

  @Synchronized
  fun mark(raw: String?): Boolean {
    val phase = AndroidStartupPhases.sanitize(raw) ?: return false
    return persistPhase(phase)
  }
}

internal data class RuntimeDiagnosticSnapshot(
  val codecContext: String? = null,
  val channelCount: Int? = null,
  val sampleRate: Int? = null,
  val selectedDecoder: String? = null,
  val passthroughEnabled: Boolean? = null,
  val downmixEnabled: Boolean? = null,
  val normalizationEnabled: Boolean? = null,
  val uiState: String? = null
)

internal object AndroidRuntimeDiagnostics {
  const val UI_STARTUP = "startup"
  const val UI_AUTHENTICATION = "authentication"
  const val UI_MAIN_SCREEN = "main_screen"
  const val UI_PLAYER = "player"
  const val UI_PLAYER_DISPOSED = "player_disposed"

  private const val PREFERENCES_NAME = "plezy_runtime_diagnostics"
  private const val KEY_CODEC_CONTEXT = "codec_context"
  private const val KEY_CHANNEL_COUNT = "channel_count"
  private const val KEY_SAMPLE_RATE = "sample_rate"
  private const val KEY_SELECTED_DECODER = "selected_decoder"
  private const val KEY_PASSTHROUGH_ENABLED = "passthrough_enabled"
  private const val KEY_DOWNMIX_ENABLED = "downmix_enabled"
  private const val KEY_NORMALIZATION_ENABLED = "normalization_enabled"
  private const val KEY_UI_STATE = "ui_state"
  private val allowedCodecContexts = setOf(
    "audio:aac",
    "audio:ac3",
    "audio:eac3",
    "audio:dts",
    "audio:truehd",
    "audio:flac",
    "audio:pcm",
    "audio:other",
    "video:dolby_vision",
    "video:hevc",
    "video:avc",
    "video:other"
  )
  private val allowedUiStates = setOf(
    UI_STARTUP,
    UI_AUTHENTICATION,
    UI_MAIN_SCREEN,
    UI_PLAYER,
    UI_PLAYER_DISPOSED
  )
  private val decoderNamePattern = Regex("[A-Za-z0-9_.:-]{1,96}")
  private val executor by lazy {
    Executors.newSingleThreadExecutor { task ->
      Thread(task, "plezy-runtime-diagnostics").apply { isDaemon = true }
    }
  }

  fun codecContextForMime(mimeType: String?): String? {
    val normalized = mimeType?.lowercase() ?: return null
    return when (normalized) {
      "audio/mp4a-latm" -> "audio:aac"
      "audio/ac3" -> "audio:ac3"
      "audio/eac3", "audio/eac3-joc" -> "audio:eac3"
      "audio/vnd.dts", "audio/vnd.dts.hd" -> "audio:dts"
      "audio/true-hd" -> "audio:truehd"
      "audio/flac" -> "audio:flac"
      "audio/raw" -> "audio:pcm"
      "video/dolby-vision" -> "video:dolby_vision"
      "video/hevc" -> "video:hevc"
      "video/avc" -> "video:avc"
      else -> when {
        normalized.startsWith("audio/") -> "audio:other"
        normalized.startsWith("video/") -> "video:other"
        else -> null
      }
    }
  }

  fun sanitizeDecoderName(raw: String?): String? {
    if (raw == null) return null
    return raw.takeIf(decoderNamePattern::matches) ?: "unknown"
  }

  fun sanitizeUiState(raw: String?): String? = raw?.takeIf(allowedUiStates::contains)
  fun sanitize(snapshot: RuntimeDiagnosticSnapshot): RuntimeDiagnosticSnapshot = RuntimeDiagnosticSnapshot(
    codecContext = snapshot.codecContext?.takeIf(allowedCodecContexts::contains),
    channelCount = snapshot.channelCount?.takeIf { it in 1..32 },
    sampleRate = snapshot.sampleRate?.takeIf { it in 1..768_000 },
    selectedDecoder = sanitizeDecoderName(snapshot.selectedDecoder),
    passthroughEnabled = snapshot.passthroughEnabled,
    downmixEnabled = snapshot.downmixEnabled,
    normalizationEnabled = snapshot.normalizationEnabled,
    uiState = sanitizeUiState(snapshot.uiState)
  )

  fun update(
    context: Context,
    codecContext: String? = null,
    channelCount: Int? = null,
    sampleRate: Int? = null,
    selectedDecoder: String? = null,
    passthroughEnabled: Boolean? = null,
    downmixEnabled: Boolean? = null,
    normalizationEnabled: Boolean? = null,
    uiState: String? = null
  ) {
    val safeCodecContext = codecContext?.takeIf(allowedCodecContexts::contains)
    val safeChannelCount = channelCount?.takeIf { it in 1..32 }
    val safeSampleRate = sampleRate?.takeIf { it in 1..768_000 }
    val safeDecoder = sanitizeDecoderName(selectedDecoder)
    val safeUiState = sanitizeUiState(uiState)
    val applicationContext = context.applicationContext
    executor.execute {
      runCatching {
        applicationContext.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE).edit().apply {
          safeCodecContext?.let { putString(KEY_CODEC_CONTEXT, it) }
          safeChannelCount?.let { putInt(KEY_CHANNEL_COUNT, it) }
          safeSampleRate?.let { putInt(KEY_SAMPLE_RATE, it) }
          safeDecoder?.let { putString(KEY_SELECTED_DECODER, it) }
          passthroughEnabled?.let { putBoolean(KEY_PASSTHROUGH_ENABLED, it) }
          downmixEnabled?.let { putBoolean(KEY_DOWNMIX_ENABLED, it) }
          normalizationEnabled?.let { putBoolean(KEY_NORMALIZATION_ENABLED, it) }
          safeUiState?.let { putString(KEY_UI_STATE, it) }
        }.commit()
      }
    }
  }

  fun clearPlayback(context: Context) {
    val applicationContext = context.applicationContext
    executor.execute {
      runCatching {
        applicationContext.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE).edit()
          .remove(KEY_CODEC_CONTEXT)
          .remove(KEY_CHANNEL_COUNT)
          .remove(KEY_SAMPLE_RATE)
          .remove(KEY_SELECTED_DECODER)
          .remove(KEY_PASSTHROUGH_ENABLED)
          .remove(KEY_DOWNMIX_ENABLED)
          .remove(KEY_NORMALIZATION_ENABLED)
          .putString(KEY_UI_STATE, UI_PLAYER_DISPOSED)
          .commit()
      }
    }
  }

  fun read(context: Context): RuntimeDiagnosticSnapshot {
    val preferences = context.applicationContext.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE)
    return RuntimeDiagnosticSnapshot(
      codecContext = runCatching { preferences.getString(KEY_CODEC_CONTEXT, null) }
        .getOrNull()
        ?.takeIf(allowedCodecContexts::contains),
      channelCount = runCatching { preferences.getInt(KEY_CHANNEL_COUNT, -1) }.getOrNull()?.takeIf { it in 1..32 },
      sampleRate = runCatching { preferences.getInt(KEY_SAMPLE_RATE, -1) }.getOrNull()?.takeIf { it in 1..768_000 },
      selectedDecoder = sanitizeDecoderName(runCatching { preferences.getString(KEY_SELECTED_DECODER, null) }.getOrNull()),
      passthroughEnabled = readBoolean(preferences, KEY_PASSTHROUGH_ENABLED),
      downmixEnabled = readBoolean(preferences, KEY_DOWNMIX_ENABLED),
      normalizationEnabled = readBoolean(preferences, KEY_NORMALIZATION_ENABLED),
      uiState = sanitizeUiState(runCatching { preferences.getString(KEY_UI_STATE, null) }.getOrNull())
    )
  }

  private fun readBoolean(preferences: SharedPreferences, key: String): Boolean? {
    if (!preferences.contains(key)) return null
    return runCatching { preferences.getBoolean(key, false) }.getOrNull()
  }
}

internal data class PreviousExitReport(
  val reason: String,
  val status: Int,
  val importance: Int,
  val timestamp: Long,
  val deviceModel: String,
  val apiLevel: Int,
  val abi: String,
  val lowRam: Boolean,
  val startupPhase: String?,
  val runtime: RuntimeDiagnosticSnapshot,
  val dedupeKey: String
) {
  fun toMap(): Map<String, Any> = buildMap {
    put("reason", reason)
    put("status", status)
    put("importance", importance)
    put("timestamp", timestamp)
    put("deviceModel", deviceModel)
    put("apiLevel", apiLevel)
    put("abi", abi)
    put("lowRam", lowRam)
    startupPhase?.let { put("startupPhase", it) }
    runtime.codecContext?.let { put("codecContext", it) }
    runtime.channelCount?.let { put("channelCount", it) }
    runtime.sampleRate?.let { put("sampleRate", it) }
    runtime.selectedDecoder?.let { put("selectedDecoder", it) }
    runtime.passthroughEnabled?.let { put("passthroughEnabled", it) }
    runtime.downmixEnabled?.let { put("downmixEnabled", it) }
    runtime.normalizationEnabled?.let { put("normalizationEnabled", it) }
    runtime.uiState?.let { put("uiState", it) }
  }
}

internal object AndroidExitReportMapper {
  private const val REASON_LOW_MEMORY = 3
  private const val REASON_CRASH = 4
  private const val REASON_CRASH_NATIVE = 5
  private const val REASON_ANR = 6
  private const val REASON_USER_REQUESTED = 10
  private const val REASON_USER_STOPPED = 11
  private const val MAX_DEVICE_MODEL_LENGTH = 80
  private val supportedAbis = setOf("arm64-v8a", "armeabi-v7a", "x86_64", "x86")

  fun map(
    record: HistoricalExitRecord,
    deviceModel: String,
    apiLevel: Int,
    abi: String,
    lowRam: Boolean,
    startupPhase: String? = null,
    runtime: RuntimeDiagnosticSnapshot = RuntimeDiagnosticSnapshot()
  ): PreviousExitReport {
    val dedupeKey = sha256(
      listOf(
        record.timestamp.toString(),
        record.reason.toString(),
        record.status.toString(),
        record.importance.toString()
      )
    )
    return PreviousExitReport(
      reason = mapReason(record.reason),
      status = record.status,
      importance = record.importance,
      timestamp = record.timestamp,
      deviceModel = sanitizeDeviceModel(deviceModel),
      apiLevel = apiLevel,
      abi = abi.takeIf(supportedAbis::contains) ?: "unknown",
      lowRam = lowRam,
      startupPhase = AndroidStartupPhases.sanitize(startupPhase),
      runtime = AndroidRuntimeDiagnostics.sanitize(runtime),
      dedupeKey = dedupeKey
    )
  }

  fun mapReason(reason: Int): String = when (reason) {
    REASON_CRASH -> "crash"
    REASON_CRASH_NATIVE -> "native_crash"
    REASON_ANR -> "anr"
    REASON_LOW_MEMORY -> "low_memory"
    REASON_USER_REQUESTED, REASON_USER_STOPPED -> "user_requested"
    else -> "other"
  }

  fun sanitizeDeviceModel(raw: String): String {
    val sanitized = buildString(raw.length.coerceAtMost(MAX_DEVICE_MODEL_LENGTH)) {
      var pendingSpace = false
      raw.forEach { character ->
        if (length >= MAX_DEVICE_MODEL_LENGTH) return@forEach
        if (character.isWhitespace() || Character.isISOControl(character)) {
          pendingSpace = isNotEmpty()
        } else {
          if (pendingSpace && length < MAX_DEVICE_MODEL_LENGTH) append(' ')
          if (length < MAX_DEVICE_MODEL_LENGTH) append(character)
          pendingSpace = false
        }
      }
    }.trim()
    return sanitized.ifEmpty { "unknown" }
  }

  private fun sha256(fields: List<String>): String {
    val digest = MessageDigest.getInstance("SHA-256")
    fields.forEach { field ->
      digest.update(field.length.toString().toByteArray(Charsets.UTF_8))
      digest.update(':'.code.toByte())
      digest.update(field.toByteArray(Charsets.UTF_8))
      digest.update(';'.code.toByte())
    }
    return digest.digest().joinToString("") { byte ->
      (byte.toInt() and 0xff).toString(16).padStart(2, '0')
    }
  }
}

internal class PreviousExitReportStore(
  private val readDedupeKey: () -> String?,
  private val persistDedupeKey: (String) -> Boolean
) {
  @Synchronized
  fun takeIfNew(report: PreviousExitReport): Map<String, Any>? {
    if (readDedupeKey() == report.dedupeKey) return null
    if (!persistDedupeKey(report.dedupeKey)) return null
    return report.toMap()
  }
}

File: lib/utils/android_exit_diagnostics.dart
import 'dart:async';
import 'dart:io';

import 'app_logger.dart';
import 'device_channel.dart';

enum AndroidStartupPhase {
  nativeOnCreate('native_on_create'),
  dartMain('dart_main'),
  runApp('runApp'),
  firstFrame('first_frame'),
  databaseOpenStarted('database_open_started'),
  databaseReady('database_ready'),
  credentialsLoaded('credentials_loaded'),
  bindingStarted('binding_started'),
  bindingSettled('binding_settled'),
  mainScreen('main_screen');

  const AndroidStartupPhase(this.id);

  final String id;
}

enum AndroidUiState {
  mainScreen('main_screen'),
  player('player');

  const AndroidUiState(this.id);

  final String id;
}

/// Best-effort bridge for the newest Android 11+ historical process exit.
abstract final class AndroidExitDiagnostics {
  static const _allowedReasons = {'crash', 'native_crash', 'anr', 'low_memory', 'user_requested', 'other'};
  static const _allowedAbis = {'arm64-v8a', 'armeabi-v7a', 'x86_64', 'x86', 'unknown'};
  static const _allowedCodecContexts = {
    'audio:aac',
    'audio:ac3',
    'audio:eac3',
    'audio:dts',
    'audio:truehd',
    'audio:flac',
    'audio:pcm',
    'audio:other',
    'video:dolby_vision',
    'video:hevc',
    'video:avc',
    'video:other',
  };
  static const _allowedUiStates = {'startup', 'authentication', 'main_screen', 'player', 'player_disposed'};
  static const _allowedStartupPhases = {
    'native_on_create',
    'dart_main',
    'runApp',
    'first_frame',
    'database_open_started',
    'database_ready',
    'credentials_loaded',
    'binding_started',
    'binding_settled',
    'main_screen',
  };
  static final _decoderNamePattern = RegExp(r'^[A-Za-z0-9_.:-]{1,96}$');
  static final _startupWatch = Stopwatch()..start();
  static var _nativeOnCreateRecorded = false;
  static var _lastElapsedMs = 0;

  /// Persists and records one fixed, privacy-safe startup phase.
  static void markStartupPhase(AndroidStartupPhase phase) {
    try {
      if (Platform.isAndroid && !_nativeOnCreateRecorded) {
        _nativeOnCreateRecorded = true;
        _recordPhase(AndroidStartupPhase.nativeOnCreate.id, 0);
      }
      final measuredMs = _startupWatch.elapsedMilliseconds;
      final elapsedMs = measuredMs < _lastElapsedMs ? _lastElapsedMs : measuredMs;
      _lastElapsedMs = elapsedMs;
      _recordPhase(phase.id, elapsedMs);
      if (Platform.isAndroid) {
        unawaited(_persistStartupPhase(phase.id));
      }
    } catch (_) {
      // Startup diagnostics must never affect the startup path.
    }
  }

  static void _recordPhase(String phase, int elapsedMs) {
    try {
      appLogger.i('Startup phase: phase=$phase elapsedMs=$elapsedMs');
    } catch (_) {
      // Local logging is best-effort.
    }
  }

  static Future<void> _persistStartupPhase(String phase) async {
    try {
      await deviceChannel.invokeMethod<bool>('setStartupPhase', phase);
    } catch (_) {
      // Native phase persistence is best-effort.
    }
  }

  static Future<void> markUiState(AndroidUiState state) async {
    if (!Platform.isAndroid) return;
    try {
      await deviceChannel.invokeMethod<bool>('setRuntimeUiState', state.id);
    } catch (_) {
      // Runtime diagnostics are best-effort and must never affect navigation.
    }
  }

  /// Records a native-sanitized previous-exit report in the local log.
  ///
  /// Native persistence makes this one-shot across launches. Every failure is
  /// intentionally contained because historical diagnostics must not affect
  /// startup.
  static Future<void> logPreviousExit() async {
    if (!Platform.isAndroid) return;
    try {
      final raw = await deviceChannel.invokeMapMethod<String, Object?>('getPreviousExit');
      final report = _validate(raw);
      if (report == null) return;

      appLogger.w(
        'Previous Android application exit: '
        'reason=${report['reason']} status=${report['status']} '
        'importance=${report['importance']} timestamp=${report['timestamp']} '
        'deviceModel=${report['deviceModel']} apiLevel=${report['apiLevel']} '
        'abi=${report['abi']} lowRam=${report['lowRam']} '
        'startupPhase=${report['startupPhase'] ?? 'omitted'} '
        'codecContext=${report['codecContext'] ?? 'omitted'} '
        'channels=${report['channelCount'] ?? 'omitted'} sampleRate=${report['sampleRate'] ?? 'omitted'} '
        'decoder=${report['selectedDecoder'] ?? 'omitted'} '
        'passthrough=${report['passthroughEnabled'] ?? 'omitted'} '
        'downmix=${report['downmixEnabled'] ?? 'omitted'} '
        'normalization=${report['normalizationEnabled'] ?? 'omitted'} '
        'uiState=${report['uiState'] ?? 'omitted'}',
      );
    } catch (_) {
      // Historical diagnostics are best-effort and must never escape startup.
    }
  }

  static Map<String, Object?>? _validate(Map<String, Object?>? raw) {
    if (raw == null) return null;
    final reason = raw['reason'];
    final status = raw['status'];
    final importance = raw['importance'];
    final timestamp = raw['timestamp'];
    final deviceModel = raw['deviceModel'];
    final apiLevel = raw['apiLevel'];
    final abi = raw['abi'];
    final lowRam = raw['lowRam'];
    final startupPhase = raw['startupPhase'];
    final codecContext = raw['codecContext'];
    final channelCount = raw['channelCount'];
    final sampleRate = raw['sampleRate'];
    final selectedDecoder = raw['selectedDecoder'];
    final passthroughEnabled = raw['passthroughEnabled'];
    final downmixEnabled = raw['downmixEnabled'];
    final normalizationEnabled = raw['normalizationEnabled'];
    final uiState = raw['uiState'];
    if (reason is! String || !_allowedReasons.contains(reason)) return null;
    if (status is! int || importance is! int || timestamp is! int) return null;
    if (deviceModel is! String ||
        deviceModel.isEmpty ||
        deviceModel.length > 80 ||
        deviceModel.runes.any(_isControlCharacter)) {
      return null;
    }
    if (apiLevel is! int || apiLevel < 1 || abi is! String || !_allowedAbis.contains(abi) || lowRam is! bool) {
      return null;
    }
    if (startupPhase != null && (startupPhase is! String || !_allowedStartupPhases.contains(startupPhase))) {
      return null;
    }
    if (codecContext != null && (codecContext is! String || !_allowedCodecContexts.contains(codecContext))) return null;
    if (channelCount != null && (channelCount is! int || channelCount < 1 || channelCount > 32)) return null;
    if (sampleRate != null && (sampleRate is! int || sampleRate < 1 || sampleRate > 768000)) return null;
    if (selectedDecoder != null && (selectedDecoder is! String || !_decoderNamePattern.hasMatch(selectedDecoder))) {
      return null;
    }
    if (passthroughEnabled != null && passthroughEnabled is! bool) return null;
    if (downmixEnabled != null && downmixEnabled is! bool) return null;
    if (normalizationEnabled != null && normalizationEnabled is! bool) return null;
    if (uiState != null && (uiState is! String || !_allowedUiStates.contains(uiState))) return null;
    return <String, Object?>{
      'reason': reason,
      'status': status,
      'importance': importance,
      'timestamp': timestamp,
      'deviceModel': deviceModel,
      'apiLevel': apiLevel,
      'abi': abi,
      'lowRam': lowRam,
      if (startupPhase case final String safeStartupPhase) 'startupPhase': safeStartupPhase,
      if (codecContext case final String safeCodecContext) 'codecContext': safeCodecContext,
      if (channelCount case final int safeChannelCount) 'channelCount': safeChannelCount,
      if (sampleRate case final int safeSampleRate) 'sampleRate': safeSampleRate,
      if (selectedDecoder case final String safeSelectedDecoder) 'selectedDecoder': safeSelectedDecoder,
      if (passthroughEnabled case final bool safePassthroughEnabled) 'passthroughEnabled': safePassthroughEnabled,
      if (downmixEnabled case final bool safeDownmixEnabled) 'downmixEnabled': safeDownmixEnabled,
      if (normalizationEnabled case final bool safeNormalizationEnabled)
        'normalizationEnabled': safeNormalizationEnabled,
      if (uiState case final String safeUiState) 'uiState': safeUiState,
    };
  }

  static bool _isControlCharacter(int rune) => rune < 0x20 || rune == 0x7f;
}
